### DISCUSSION TITLE
(Off Topic) ¿Cuantas líneas tiene mi código?

### DISCUSSION THREAD

**Discussion by IoachimusRoderici** [INGCOM-UNRN-P1/cursada-2023#194] [Mostrá algo que hayas hecho]
¿Alguna vez te preguntaste cuantas líneas de código tenés escritas en un proyecto? Yo si.

Así que escribí [un programa](https://github.com/IoachimusRoderici/MyLib/tree/main/cuantas_lineas) que las cuenta.
Y que te muestra una tablita con:
- La cantidad de líneas de código
- La cantidad de líneas de comentarios
- La cantidad de líneas vacías

Por ejemplo:
```bash
$ cuantas_lineas.py tp3-2023-IoachimusRoderici
En el directorio 'tp3-2023-IoachimusRoderici':
╒══════════════╤══════════╤═══════════════╤══════════╤═════════╤═══════════════╕
│ Archivo      │   Código │   Comentarios │   Vacías │   Total │   % de código │
╞══════════════╪══════════╪═══════════════╪══════════╪═════════╪═══════════════╡
│ ejercicio1.c │      141 │            70 │       43 │     250 │          56.4 │
│ ejercicio2.c │      203 │           189 │       63 │     452 │          44.9 │
│ plantilla.c  │        4 │             4 │        5 │      13 │          30.8 │
│ TOTAL        │      348 │           263 │      111 │     715 │          48.7 │
╘══════════════╧══════════╧═══════════════╧══════════╧═════════╧═══════════════╛
```

Disclaimer: Probablemente no funcione bien en Windows.

**Comment by martinvilu** [comment]
¿Contás las líneas en las que hay código y comentario?

Por ejemplo:
```C
int i = 0; // declaramos e inicializamos un contador
```

**Reply by IoachimusRoderici** [reply]
Si. Como dice [el docstring](https://github.com/IoachimusRoderici/MyLib/blob/f543b88c25ece7a3f35c4ccf6974ea6509b04c60/cuantas_lineas/cuantas_lineas.py#L24C1-L28):

> Se considera código  a  todo el texto que no esté dentro de un comentario.  Las
líneas que tienen  código seguido por un comentario  suman a ambos  contadores,
así que es  posible que la suma de líneas  de código y líneas de comentario sea
mayor al número  total de líneas.  Las líneas que abren o cierran un comentario
pero no le agregan contenido no suman a la cuenta de comentarios.

**Reply by martinvilu** [reply]
Muy piola

¿Y si hay una mezcla de comentario de una línea y multilínea en una sola?

Estoy desde hace algún tiempo tratando de juntar tiempo para automatizar la revisión de estilo con Python y que eso se revise mecánicamente. (https://github.com/eliben/pycparser)

**Reply by IoachimusRoderici** [reply]
> ¿Y si hay una mezcla de comentario de una línea y multilínea en una sola?

Hice varias [pruebas](https://github.com/IoachimusRoderici/MyLib/blob/main/cuantas_lineas/test_contenido_de_linea.py) mezclando comentarios de ambos tipos en una misma línea y todas salieron bien. Por ejemplo:

```C
// *//**/
/*
// *//**/
/* // */
int i; // */
```

**Reply by IoachimusRoderici** [reply]
El problema que sí hay es cuando aparecen delimitadores de comentarios adentro de string literals. 
Esa es una limitación de mi diseño que no me dieron ganas de resolver.

**Comment by martinvilu** [comment]
> Disclaimer: Probablemente no funcione bien en Windows.

¿Por que?

**Reply by IoachimusRoderici** [reply]
Mi código considera ocultos a los archivos y directorios que empiezan con `.`.
Desconozco cómo maneja Windows los archivos ocultos pero creo que es distinto que en Linux.

Exceptuando ese detalle *creo* que andaría bien, pero como no lo escribí con Windows en mente me pareció que no estaba de más la aclaración.

**Reply by martinvilu** [reply]
Funcionaria igual, lo oculto en Windows es por un atributo más que por el nombre.

Esto significa que procesaría un archivo oculto pero con nombre 'normal'

**Reply by IoachimusRoderici** [reply]
Claro, eso me temía.

Entonces si corrés el programa en Windows en un directorio que tiene carpetas `.git`, `.vscode`, etc. probablemente te incluya un montón de archivos que no querías contar (o no, no sé que hay adentro de esas carpetas).

**Reply by martinvilu** [reply]
Estando en estilo unix, va a funcionar igual

**Reply by IoachimusRoderici** [reply]
Ah claro 🤦. Entonces da igual.
